flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/calcite/FlinkTypeFactory.scala

```diff
@@ -379,14 +380,25 @@ object FlinkTypeFactory {
       case BIGINT => new BigIntType()
       case FLOAT => new FloatType()
       case DOUBLE => new DoubleType()
-      case VARCHAR | CHAR =>
-        // TODO we use VarCharType to support sql CHAR, VarCharType don't support 0 length
-        new VarCharType(
-          if (relDataType.getPrecision == 0) VarCharType.MAX_LENGTH else relDataType.getPrecision)
-      case VARBINARY | BINARY =>
-        // TODO we use VarBinaryType to support sql BINARY, VarBinaryType don't support 0 length
-        new VarBinaryType(
-          if (relDataType.getPrecision == 0) VarBinaryType.MAX_LENGTH else relDataType.getPrecision)
+      case CHAR =>
+        if (relDataType.getPrecision == 0) {
+          CharType.ofEmptyLiteral
+        } else {
+          new CharType(relDataType.getPrecision)
+        }
+      case VARCHAR =>
+        // Calcite will return a varchar RelDataType with precision 0, for example,
+        // SUBSTR('', 2, -1), Calcite infers that the return value must not exist
+        // (in this case, null will be returned when executed in Blink runner),
+        // so it infers VARCHAR(0).  But our `VarcharType` not allow precision 0.
+        new VarCharType(Math.max(1, relDataType.getPrecision))
```

Comment:
Consider sql: `SELECT trim(LEADING  from '') FROM T`
`SqlTrimFunction` use `ReturnTypes.cascade(ReturnTypes.ARG2, SqlTypeTransforms.TO_NULLABLE, SqlTypeTransforms.TO_VARYING)` to infer return type.
`TO_VARYING` just convert Char to VarChar and use previous precision which maybe 0.
Now in Calcite, there is no limit to zero precision.

We need introduce an `ofEmptyLiteral()` to the variable length types as well.

Reply:
After https://github.com/apache/flink/pull/8730 , it is ready now.